dotnet/src/SemanticKernel.Abstractions/Orchestration/FunctionResult.cs

```diff
@@ -80,4 +81,15 @@ public FunctionResult(string functionName, string pluginName, SKContext context,
 
         throw new InvalidCastException($"Cannot cast {this.Value.GetType()} to {typeof(T)}");
     }
+
+    /// <summary>
+    /// Populates metadata with variables from <see cref="SKContext"/>.
+    /// </summary>
+    internal void PopulateMetadata()
+    {
+        foreach (var variable in this.Context.Variables)
+        {
+            this.Metadata[variable.Key] = variable.Value;
```

Comment:
I'm wondering, if it will be possible to use `Metadata` dictionary only for fields that are required to expose, but not all the variables from `Context`. Otherwise, we can just make `Context` or `Context.Variables` field public. But main idea here was to expose only what is needed for end-user and hide unnecessary context information. Because if context information is opened, the context is used in the wrong way for workarounds and hacks :)

In this scenario, we can add to metadata key `StepwisePlanner` with value, which could be complex object, including properties `StepCount`, `FunctionCount` and `Iterations`. These properties could be set somewhere in `StepwisePlanner` logic before returning `FunctionResult`, if that is possible.

Reply:
Yeah this is fair. Will move to `Plan` object where `Outputs` can be directly referenced.